executor/README.md

```diff
@@ -0,0 +1,270 @@
+
+In cases where only one copy of a certain job should be running, before and after commands could be used to acquire and release a lock. Assuming you've written some scripts for managing your lock, you could execute them by passing the additional data to the cook scheduler's job creation API:
+
+```json
+{
+	"command": "my-actual-command",
+	"before_commands": [{
+		"value": "./acquire-zk-lock.sh my-service"
+		"guard": true
+	}],
+	"after_commands":[{
+		"value": "./release-zk-lock.sh my-service"
+	}],
+	"uris": [{
+		"value": "http://my-file-server/release-zk-lock.sh",
+		"executable": true
+  	},{
+  		"value": "http://my-file-server/acquire-zk-lock.sh",
+  		"executable": true
+  	}]
+}
+```
+
+The `guard` flag is used to prevent execution of later commands if we can't acquire the lock (in this case, `acquire-zk-lock.sh` script will return a non-zero exit code). Note that the scripts for managing the lock could be completely reusable, and that the primary command did not need to change.
+
+#### Monitoring a running command
+
+Perhaps you'd like to tail a log created by your primary command. You could use an asynchronous before command to do so. It will be started before your primary command, and continue running in the background until the primary command (and all other synchronous commands) finish.
+
+```json
+{
+	"command": "my-actual-command",
+	"before_commands": [{
+		"value": "./my-monitoring-command",
+		"async": true
+	}]
+}
+```
+
+#### Dynamically setting environment variables
+
+In order to dynamically set environment variables for later commands, you can use the executor HTTP API from a before command.
+
+```json
+{
+	"command": "my-actual-command",
+	"before_commands": [{
+		"value": "./set-env-vars.sh"
+	}]
+}
+```
+
+The before command would need to executor something like the following:
+
+```bash
+curl -XPATCH \
+	-d '{"env": {"MY_ENV_VAR": "/hello/world", "MY_OLD_VAR": null}}' \
+	$EXECUTOR_ENDPOINT
+```
+
+This would set `MY_ENV_VAR` for your primary command (and all commands after `set-env-vars.sh`.
+
+#### Update progress and custom failure messages
+
+In order to update progress or add a custom failure message, you can use the executor HTTP API from any command, with something like the following:
+
+```bash
+curl -XPATCH \
+	-d '{"progress": 100, "message": "a custom failure message"}' \
+	$EXECUTOR_ENDPOINT
+```
+
+#### Sending an email on completion
+
+In order to send an email once your primary command has finished, you can use an after command:
+
+```json
+{
+	"after_commands": [{
+		"value": "mail -s 'your job failed, sorry' bob@hello.com"
+	}]
+}
+```
+
+You could also take advantage of the read API to check the exit code of the primary command in order to only send an email on error. Here's an example script using `bash` and `jq`:
+
+```bash
+exit_code=$(curl $EXECUTOR_ENDPOINT | jq '.exit_code')
+
+if [[ "$exit_code" != "0" ]]; then
+    mail -s 'your job failed, sorry!' bob@hello.com
+fi
+```
+
```

Comment:
Would be very useful to have an endpoint to kill the main process. For example, consider the case in which a before_command is a progress monitor, which is supposed to kill the main command when no progress is made for x amount of time. Then it could just hit such endpoint. @wyegelwel what do you think?